### ISSUE TITLE
False positive PLE0302 with special methods as properties.

### ISSUE THREAD

**Issue by mikeshardmind** [astral-sh/ruff#15572]
Example code that works at runtime and which type checkers do understand

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgEEcBPLGAcSSyRANWoBoo7Tt14gBAYQKJiyCdTABXDJwBQoSFFSMK6LAHMylarTpZGK8lRqxsMBSBAdUeYEvtIcUAp5jAVKrQokegBRADkAFRCAJSgAXihJaSJkAG1UgF05cCVdJFTTRgFC4qYWIS4eagKzKAAfKDDsJAzWgO1guhCADQBJCPjEqXgZfNTA-IAhbyQQgA8YJAoMbAz6xuaBaZxZhaWVrHWmzgEwIgArFFQsxPlczhqi%2BjMBIjAENpUYEZxPOnEIt1UgBNDIALhUUChUAAJkhgFAAPqI3ToVDIgAUO3gwAEejB9DKbA4lVEIJuwAJvjwADE3I5UscWgBKKAAWgAfBtOBDoXyoNjgHgDAk9JD%2BVDBS5Bn5xVDvMxSHCEciCE4uJjBazOYTFRURNVQbyJVBHKh7IcpWK5V4ibD4UjEWq5mjMTa%2BYK%2BO7oUgFoiJgSJqltrtFstMFg1g0mV6TVBfTBEQA3KQKJAEkPzMMHI6bb1Q1DcRZEAgwADWBLOlzQseh2q5bwQxol6ARCeTqeC6E8WDAtCZzbjAqQOJceB26IT%2BwjGPbKfgaeZ-jjBAA7gQ0cPR3pCN8wDsMUu42aLVAaVIdv4VAqWPaVQQMWg5gT-oDHhl61B4N3UO-BzfSFXNEAAtYFQOZbVgQc%2BRPEBDlSJNbUVKBgGoKBEN0WAMhtWD4Ow-xvm8Tx5hQJQkAAdW4CgghAQdlUdVF0URLER1xJEYBfAEgUKD9oOhKVEVIBJBOXaF6NVddXWYrV2S5fUqhAR5iheL8f3fcF81NJBzTgpEHwEmBmTwCSNyYw9RKhAABChwBorQbXEp11RATVWOZPioVwrchUE4ynKwVANURG0bWs2yuHsvlHOdKSWJxdzNK8gy-JipiLKQ296OAbSYGA5o4uAT9ClzHlNIAqAgNQUDBUg-FNL5crUJAdCpDIQ46qHE0vPnRLtNPJkgA)

```python
from collections.abc import AsyncGenerator, Generator, Callable, Coroutine
from typing import Any
import concurrent.futures as cf

type AENTER = Callable[[], Coroutine[Any, Any, AsyncGenerator[Any | None]]]
type AEXIT = Callable[[type[BaseException] | None, BaseException | None, object], Coroutine[Any, Any, bool]]

class ACTX[Y]:
    def __init__(self, g: AsyncGenerator[Y], f: cf.Future[None]) -> None:
        self.g = g
        self.f = f

    async def __aenter__(self) -> AsyncGenerator[Y]:
        return self.g

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_value: BaseException | None,
        traceback: object,
    ) -> bool:
        if exc_value is not None:
            self.f.set_exception(exc_value)

        await self.g.aclose()
        return False


async def _a(ctx: ACTX[Any]) -> list[Any]:
    async with ctx as c:
        return [v async for v in c]
    return []


class ExecuteWrapper:
    def __init__(self, _c: ACTX[Any]):
        self._c = _c

    def __await__(self) -> Generator[Any, Any, list[Any]]:
        return _a(self._c).__await__()

    @property
    def __aenter__(self):
        return self._c.__aenter__
    
    @property
    def __aexit__(self):
        return self._c.__aexit__

    async def fetchone(self) -> Any | None:
        async with self as g:
            async for val in g:
                return val
        return None
```

Rewriting `ExecuteWrapper` as:

```py
from operator import attrgetter

class ExecuteWrapper:
    def __init__(self, _c: ACTX[Any]):
        self._c = _c

    def __await__(self) -> Generator[Any, Any, list[Any]]:
        return _a(self._c).__await__()

    __aenter__: AENTER = property(attrgetter("_c.__aenter__"))  # pyright: ignore[reportAssignmentType]
    __aexit__: AEXIT = property(attrgetter("_c.__aexit__"))  # pyright: ignore[reportAssignmentType]

    async def fetchone(self) -> Any | None:
        async with self as g:
            async for val in g:
                return val
        return None
```
for contrast has better runtime behavior (slightly, anyhow), but unfortunately, properties are not generic in the type system, so this ends up not type-checked, though ruff has no issue with this definition.

**Comment by dylwil3** [comment]
I think it seems reasonable to skip the check when there's a property decorator (or maybe _any_ decorator).

**Comment by charliermarsh** [comment]
I assume I'm lacking understanding. Can `ExecuteWrapper` not use methods for this?

```python
    async def __aenter__(self):
        return await self._c.__aenter__()

    async def __aexit__(self, exc_type, exc_value, traceback):
        return await self._c.__aexit__(exc_type, exc_value, traceback)
```

**Comment by mikeshardmind** [comment]
Using properties here has multiple benefits, from better automatic type inference to better runtime behavior. With using methods, this would be multiple awaits for a simple wrapper or require `yield from self._c.__aexit__(exc_type, exc_value, traceback).__await__()` which is much harder for human reviewer to reason about.

**Comment by mikeshardmind** [comment]
This might be better to hold off on for now, there are enough static analysis tools that are getting subtle things wrong about dunder methods as descriptors beyond what is visible here that it may still be beneficial to nudge people to slightly different solutions even at a small runtime or legibility cost to ensure their tooling has the expected benefits.
